### PR TITLE
Protocol data types

### PR DESCRIPTION
**Pull request by repnop**

### CODE REVIEW

File: src/ber/mod.rs [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+    pub const fn new(byte: u8) -> Self {
+        Tag(byte)
+    }
+
+    pub const fn from_parts(class: Class, aspect: Aspect, tag: u8) -> Self {
+        Tag((tag & 0x1F) | ((class as u8) << 6) | ((aspect as u8) << 5))
+    }
+
+    /// Get the `Class` of the tag
+    pub fn class(self) -> Class {
```
Comment (Dylan-DPC-zz) [inline src/ber/mod.rs]: why does this consume self?

File: src/ber/mod.rs [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+            0 => Class::Universal,
+            1 => Class::Application,
+            2 => Class::ContextSpecific,
+            3 => Class::Private,
+            _ => unreachable!("matching on two bits"),
+        }
+    }
+
+    /// Get the `Aspect` of the tag
+    pub fn aspect(self) -> Aspect {
```
Comment (Dylan-DPC-zz) [inline src/ber/mod.rs]: ~same as above, may not need to consume~

File: src/ber/mod.rs [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+            _ => Err(DeserializeError::IndefiniteLength),
+        }
+    }
+}
+
+impl Serialize for Length {
+    fn serialize(&self, buffer: &mut dyn VecExt) {
+        if self.0 > 127 {
+            let ns = self.0.to_be_bytes();
+            let start = ns.iter().copied().position(|n| n != 0).unwrap_or(7);
```
Comment (Dylan-DPC-zz) [inline src/ber/mod.rs:393]: isn't this just taking the 1st index? maybe we can use `get(1).unwrap_or(7)`?

File: src/ber/mod.rs [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+impl std::ops::Add for Length {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self {
+        Self(self.0 + rhs.0)
+    }
+}
+
+impl Serialize for str {
+    fn serialize(&self, buffer: &mut dyn VecExt) {
```
Comment (Dylan-DPC-zz) [inline src/ber/mod.rs:411]: i think this could work with generics instead of dyn dispatch

File: src/ber/mod.rs [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+}
+
+/// LDAP defines a BOOLEAN `true` to be `0xFF` and any other value `false`
+impl Serialize for bool {
+    fn serialize(&self, buffer: &mut dyn VecExt) {
+        BOOL.serialize(buffer);
+        Length::new(1).serialize(buffer);
+
+        match self {
+            true => buffer.write_byte(0xFF),
```
Comment (Dylan-DPC-zz) [inline src/ber/mod.rs:608]: clippy will complain about this and tell you to use a normal `if` / `else` instead

File: src/ber/util.rs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+impl VecExt for TinyVec<[u8; 32]> {
+    fn write(&mut self, slice: &[u8]) {
+        self.extend_from_slice(slice);
+    }
+    fn write_byte(&mut self, byte: u8) {
+        self.push(byte);
+    }
+}
+
+pub trait ReadExt {
```
Comment (Dylan-DPC-zz) [inline src/ber/util.rs]: given that the error is same for all, we could use an associated type instead. not necessary but just a thought

File: src/ber/util.rs [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+        Ok(*self.get(0).ok_or(DeserializeError::BufferTooShort)?)
+    }
+
+    fn byte(&mut self) -> Result<u8, DeserializeError> {
+        let byte = *self.get(0).ok_or(DeserializeError::BufferTooShort)?;
+        *self = &self[1..];
+        Ok(byte)
+    }
+
+    fn uint(&mut self, size: Length) -> Result<u64, DeserializeError> {
```
Comment (Dylan-DPC-zz) [inline src/ber/util.rs:48]: if we are taking in Length and converting to `usize`, why not accept a `usize` directly?

File: src/ber/util.rs [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+        let mut bytes = [0; 8];
+        bytes[8 - size..][..size].copy_from_slice(&self[..size]);
+
+        *self = &self[size..];
+
+        Ok(u64::from_be_bytes(bytes))
+    }
+
+    fn int(&mut self, size: Length) -> Result<i64, DeserializeError> {
```
Comment (Dylan-DPC-zz) [inline src/ber/util.rs:65]: same as above

File: src/ber/util.rs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+        bytes[8 - size..][..size].copy_from_slice(&self[..size]);
+
+        *self = &self[size..];
+
+        Ok(i64::from_be_bytes(bytes))
+    }
+
+    fn slice(&mut self, size: Length) -> Result<&[u8], DeserializeError> {
+        let size = size.0 as usize;
```
Comment (Dylan-DPC-zz) [inline src/ber/util.rs:84]: same ^

File: src/ber/util.rs [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+            Err(DeserializeError::BadTag { expected, got: tag })
+        } else {
+            Ok(tag)
+        }
+    }
+}
+
+/// This helper function creates a premade size and fills it in after the
+/// function is complete
+pub(crate) fn serialize_sequence<F: Fn(&mut dyn VecExt)>(buffer: &mut dyn VecExt, f: F) {
```
Comment (Dylan-DPC-zz) [inline src/ber/util.rs:119]: maybe we don't need dyn dispatch here and plain ol' generics might be sufficient

File: src/ber/mod.rs (hunk 1 above)
Comment (repnop) [reply src/ber/mod.rs]: `Tag` is `Copy`

File: src/ber/mod.rs (hunk 3 above)
Comment (repnop) [reply src/ber/mod.rs:393]: No, `u64::to_be_bytes()` returns a `[u8; 8]`, and this finds the first non-zero byte in that array

File: src/ber/mod.rs (hunk 4 above)
Comment (repnop) [reply src/ber/mod.rs:411]: I don't think it really matters to be honest

File: src/ber/util.rs (hunk 7 above)
Comment (repnop) [reply src/ber/util.rs:48]: Because `usize`'s size is platform dependent

File: src/ber/util.rs (hunk 10 above)
Comment (repnop) [reply src/ber/util.rs:119]: Again, I don't really think it matters, any network latencies are going to dwarf the whole serialization/deserialization process anyway, and even then we'd need to bench to see if there's even a benefit to it

File: src/ber/mod.rs (hunk 5 above)
Comment (repnop) [reply src/ber/mod.rs:608]: I actually don't seem to get a warning about this -- I know I have in the past. Regardless, I think the match is more readable than the if else, so I'll just allow the lint in the future if necessary

File: src/ber/util.rs (hunk 6 above)
Comment (repnop) [reply src/ber/util.rs]: Not a bad idea, but pretty low priority, might do it in a future commit once the error handling stuff is actually set up

File: src/ber/mod.rs (hunk 4 above)
Comment (Dylan-DPC-zz) [reply src/ber/mod.rs:411]: it doesn't , but it could make things simpler. Honestly just a tidbit

File: src/ber/util.rs (hunk 10 above)
Comment (Dylan-DPC-zz) [reply src/ber/util.rs:119]: yes definitely. just a tidbit

File: src/ber/util.rs (hunk 7 above)
Comment (repnop) [reply src/ber/util.rs:48]: Realistically though anything greater than u16 probably will never happen, I'll give that some thought and come back to it sometime to see if the change is worth it

File: src/ber/mod.rs [hunk 11]
```diff
@@ -0,0 +1,619 @@
...
+//! | Name                   | Aspect      | Tag Number dec [hex] |
+//! |------------------------|-------------|----------------------|
+//! | BOOLEAN                | Primitive   | 1 [0x01]             |
+//! | INTEGER                | Primitive   | 2 [0x02]             |
+//! | OCTET STRING           | Primitive^  | 4 [0x04]             |
+//! | SEQUENCE / SEQUENCE OF | Constructed | 16 [0x10]            |
+//! | SET / SET OF           | Constructed | 17 [0x11]            |
+//!
+//! ^ `OCTET STRING` is defined as both primitive and constructed but the LDAP
+//! specification restricts it to the Primitive encoding only
```
Comment (amadeusine) [inline src/ber/mod.rs]: Note for discussion: we could look into support for superscript tags with docs notes and annotations. Possibly relevant is `<sup>`?

File: src/ber/mod.rs [hunk 12]
```diff
@@ -0,0 +1,619 @@
...
+//! defines boolean `TRUE` to be `FF`, so this is `FALSE`. `newSuperior` is an
+//! optional context-specific `LDAPDN`. So we check to see if we're at the end
+//! of the message yet -- nope! still have more bytes to process, then we check
+//! the tag: `80` has a bit pattern of `10|0|00000` which tells us its a
+//! context-specific, universal type with a tag number of 0, that's our type!
+//! The length and content are decoded as usual. We're still not done however,
+//! we finished decoding `ModifyDNRequest`, but we still have the `controls`
+//! field of `LDAPMessage`, which is also an optional type. Check one: have we
+//! reached the end of the packet? As a matter of fact, yes, we have! Therefore
+//! `controls` is `None` and we've finished decoding the entire packet!
```
Comment (amadeusine) [inline src/ber/mod.rs]: This section is very helpful. For someone just entering the crate, thoughts on whether we could bisect this into two or more distinct portions?

One possible position to split could be:
> So we check to see if we're at the end of the message yet -- nope!

**Review by amadeusine — APPROVED** [review]

**Comment by amadeusine** [comment]
As an accompanying note from when we talked about `crates.io`, changing `pub` exposure so `docs.rs` can give an initial render might be something to consider. Everything else looks fine so far. Great work, Wes.

File: src/ber/mod.rs (hunk 12 above)
Comment (repnop) [reply src/ber/mod.rs]: Yeah after I wrote the whole thing I wanted to try and split it up slightly more, but wasn't sure where the best place would be. Perhaps I'll even try rewording a couple things slightly so it flows better in paragraph form.

**Comment by amadeusine** [comment]
As another note, one item to consider is the handling of serialisation for data where it would make little sense to do so. Currently, the decision to segregate with enums is one potential direction we could follow. Leaving this to further discussion.

**Pull request merged**